File: 3041-maximize-consecutive-elements-in-an-array-after-modification/3041-maximize-consecutive-elements-in-an-array-after-modification.cpp
class Solution {
public:
    int maxSelectedElements(vector<int>& nums) {
        int n = nums.size();
        sort(nums.begin(), nums.end());

        unordered_map<int, int>maxFor;
        int maxSize = 0;
        for(int i = n - 1; i >= 0; --i){
            if(maxFor.find(nums[i] + 1) == maxFor.end()){
                maxFor[nums[i]] = 1;
            }
            else{
                maxFor[nums[i]] = 1 + maxFor[nums[i] + 1];
            }

            if(maxFor.find(nums[i] + 2) == maxFor.end()){
                maxFor[nums[i] + 1] = 1;
            }
            else{
                maxFor[nums[i] + 1] = 1 + maxFor[nums[i] + 2];
            }

            maxSize = max(maxSize, maxFor[nums[i]]);
            maxSize = max(maxSize, maxFor[nums[i] + 1]);
        }
        return maxSize;
    }
};

File: 3041-maximize-consecutive-elements-in-an-array-after-modification/3041-maximize-consecutive-elements-in-an-array-after-modification.java
class Solution {
    public int maxSelectedElements(int[] nums) {
        Arrays.sort(nums);
        int[] arr = new int[1000001];
        int ans = 1;
        
        for(int i : nums){
            int prev1 = arr[i-1];
            int prev2 = arr[i];
            ans = Math.max(prev1+1,Math.max(prev2+1,ans));
            //System.out.println(i+" "+ans);
            arr[i] = Math.max(arr[i],prev1+1);
            arr[i+1] = Math.max(prev2+1,arr[i+1]);
        }
        return ans;
    }
}